Combine WTF clang submodules into one where possible.

adetaylor opened the pull request:
#### b41b7e718111d895fe022a4f325b59d086be684f
<pre>
Combine WTF clang submodules into one where possible.
<a href="https://bugs.webkit.org/show_bug.cgi?id=298441">https://bugs.webkit.org/show_bug.cgi?id=298441</a>
<a href="https://rdar.apple.com/159939714">rdar://159939714</a>

Reviewed by Elliott Williams.

This commit significantly changes the modulemap made for the WTF component of
WebKit. Previously, each WTF header (with a few exceptions) ended up in its
own explicit clang module - essentially there were hundreds of small
modules. With this change, we combine all headers into a single module
except for a small number of exceptions, as outlined in
the modulemap.toml comments.

It is hoped that this demands less precision from users of WTF about which
exact header files they include.

There are two other intentional functional changes in this commit.
Previously, the outermost &quot;requires cplusplus11&quot; directive
accidentally applied even to the Platform submodule which was OK
to be included in plain C code. This has been fixed; there is no
longer such an outermost &quot;requires cplusplus11&quot; directive and
all parts of WTF which require C++ are now in a single large
submodule.

In addition, the config macro UCHAR_TYPE is no longer necessary to build
WTF properly, and has been removed.

* Source/WTF/Configurations/modulemap.toml:
* Source/WTF/Scripts/modulemap-from-tapi-filelist.py:

Canonical link: <a href="https://commits.webkit.org/299795@main">https://commits.webkit.org/299795@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6883c55f0c59e878b4a391f8fa6364ddab2987df

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120111 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39805 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30456 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126460 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72184 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/fab5e71c-3185-48e9-95c8-5c16d3068a81) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40501 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48382 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91220 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60526 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/5a96be29-a50f-4981-9ab7-35533fed63fc) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123063 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32353 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107695 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71771 "2 api tests failed or timed out") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/16dc15ce-0680-43ab-8950-87e078e18e50) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31387 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25801 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70087 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/112236 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101832 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25988 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129364 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/118627 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47031 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35677 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99835 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47397 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103883 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99678 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25323 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45135 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23157 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43662 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46893 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52599 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/147326 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46361 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37858 "Found 1 new JSC binary failure: testapi, Found 18674 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_sort2.js.default, ChakraCore.yaml/ChakraCore/test/Array/push2.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/bug215238.shrua-2.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default ... (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49708 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48045 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

adetaylor commented:
@emw-apple Do you think this is a good idea? Please review and r+ if so! (As ever, please r+ even if you have minor requests to make life easier from this timezone). But I won't land until @JonWBedard says the toolchain is updated next week.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/809e3267bff6c05fd5e59ca1a4ad08c47981a65b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50358

emw-apple commented on Source/WTF/Configurations/modulemap.toml:
```diff
@@ -1,46 +1,47 @@
-# as an umbrella directory.
-one-submodule-per-header = true
-
+discovered-headers-module = 'Cpp'
```
OK, so what this essentially means is that if a header _isn't_ assigned to a `[module.<x>]` section, it gets pulled into one overarching submodule named `wtf.Cpp`? Seems reasonable!

I wonder if calling it `default-module` is easier to explain than "discovered headers", but I don't feel strongly about it either way.

adetaylor replied:
Exactly. And happy to rename, I'll do that

emw-apple commented:
With this change, the number of WTF submodules is now fixed, which makes me wonder if we actually benefit from having the modulemap generator. Like, would it be possible to have a

```module wtf {
    requires cplusplus11
    export *
}
```

with some inner submodules for the headers that are C-compatible or need to be isolated? Or is that not possible, because of the behavior of `requires cplusplus11`, or something else?

adetaylor committed:
Combine WTF clang submodules into one where possible.
https://bugs.webkit.org/show_bug.cgi?id=298441
rdar://159939714

Reviewed by Elliott Williams.

This commit significantly changes the modulemap made for the WTF component of
WebKit. Previously, each WTF header (with a few exceptions) ended up in its
own explicit clang module - essentially there were hundreds of small
modules. With this change, we combine all headers into a single module
except for a small number of exceptions, as outlined in
the modulemap.toml comments.

It is hoped that this demands less precision from users of WTF about which
exact header files they include.

There are two other intentional functional changes in this commit.
Previously, the outermost "requires cplusplus11" directive
accidentally applied even to the Platform submodule which was OK
to be included in plain C code. This has been fixed; there is no
longer such an outermost "requires cplusplus11" directive and
all parts of WTF which require C++ are now in a single large
submodule.

In addition, the config macro UCHAR_TYPE is no longer necessary to build
WTF properly, and has been removed.

* Source/WTF/Configurations/modulemap.toml:
* Source/WTF/Scripts/modulemap-from-tapi-filelist.py:

Canonical link: https://commits.webkit.org/299795@main

adetaylor commented:
> It is possible that we could do something using `umbrella` to populate the nested C++ module, though I'm not very optimistic, but yeah I should have tried it - I will do. I'm in the middle of a few other things so it might take a few days. Would you be OK with landing this now and then I'll try to remove the script as a follow-up? I assume so since you've r+'ed it.

Follow up PR https://github.com/WebKit/WebKit/pull/51245
